src/common/raft/integration/raft_knowhere_index.cuh

```diff
@@ -454,40 +486,42 @@ struct raft_knowhere_index<IndexKind>::impl {
-                            std::numeric_limits<indexing_type>::max(), indexing_type{-1});
-        }
-
-        // For raft_ivf_flat, setting invalid IDs to 0 here will only
-        // slightly affect the recall(ground truth contain 0th vector). In order to
-        // maintain consistency in Knowhere's results and better serve the merging of results in the upper layers of
-        // Milvus, it is temporarily replaced with -1
-        if constexpr (index_kind == raft_proto::raft_index_kind::ivf_flat) {
-            thrust::replace(res.get_thrust_policy(), thrust::device_ptr<indexing_type>(device_ids.data_handle()),
-                            thrust::device_ptr<indexing_type>(device_ids.data_handle() + output_size), indexing_type{0},
-                            indexing_type{-1});
-        }
-
-        if constexpr (index_kind == raft_proto::raft_index_kind::cagra) {
-            auto tmp = raft::make_device_matrix<knowhere_indexing_type, input_indexing_type>(res, row_count, k);
-            raft::copy(res, tmp.view(), device_ids);
-            thrust::replace(res.get_thrust_policy(), thrust::device_ptr<knowhere_indexing_type>(tmp.data_handle()),
-                            thrust::device_ptr<knowhere_indexing_type>(tmp.data_handle() + output_size),
-                            knowhere_indexing_type{0x7fffffffu}, knowhere_indexing_type{-1});
-            raft::copy(res, host_ids, tmp.view());
 
+        if constexpr (index_kind == raft_proto::raft_index_kind::brute_force) {
+            if (k_tmp > k) {
+                for (auto i = 0; i < host_ids.extent(0); ++i) {
+                    raft::copy(res, raft::make_host_vector_view(host_ids.data_handle() + i * host_ids.extent(1), k),
+                               raft::make_device_vector_view(device_ids.data_handle() + i * device_ids.extent(1), k));
+                    raft::copy(
+                        res,
+                        raft::make_host_vector_view(host_distances.data_handle() + i * host_distances.extent(1), k),
+                        raft::make_device_vector_view(device_distances.data_handle() + i * device_distances.extent(1),
+                                                      k));
+                }
+            } else {
+                raft::copy(res, host_ids, device_ids);
+                raft::copy(res, host_distances, device_distances);
+            }
         } else {
             raft::copy(res, host_ids, device_ids);
+            raft::copy(res, host_distances, device_distances);
+        }
+        raft::resource::sync_stream(res);
+        auto static max_distance = std::nextafter(std::numeric_limits<data_type>::max(), 0.0f);
+        for (auto i = knowhere_indexing_type{}; i < row_count; ++i) {
+            for (auto j = knowhere_indexing_type{}; j < k; ++j) {
+                auto distance = host_distances(i, j);
+                if (distance >= max_distance || distance < 0 ||
```

Comment:
Will handling this on the host side result in poorer performance compared to GPU, especially in large-scale K?

Reply:
That's a very good question and something I mentioned in [this comment](https://github.com/zilliztech/knowhere/pull/286#discussion_r1424309842) on the PR that is really responsible for this change. The short answer is that it really depends for both batch size and k. Sometimes, the device allocation and copy is more expensive than the host-side compute; sometimes the host-side compute is more expensive.

I'm actually fine with doing it either way in the initial version. I don't think we'll have really good guidance until we do more thorough benchmarking. Happy to change it in #286 if you'd rather be conservative there.

Reply:
I just took a closer look at my initial perf testing and realizing that this has more of an impact than I thought. I've just updated #286 with [this change](https://github.com/zilliztech/knowhere/pull/286/commits/d3f459f432d1cf80b96248e3c6ef588e409b8968) that allows us to avoid the extra allocation when it's not necessary, and that gets us much better performance all the way around.

@Presburger would you mind taking a look at that change, and if you like it, I'll cherry-pick it onto #292 and #296?

Reply:
@wphicks I prefer substitution on the device side. Comparatively, there are more performance factors on the host side.

Reply:
Excellent! I'll rebase to pull in those changes. Thanks very much!